### DISCUSSION TITLE
Torch-TensorRT Dynamo Design Documentation for Developers

### DISCUSSION THREAD

**Discussion by gs-olive** [pytorch/TensorRT#2475] [Internals]
# Overview
The Torch-TensorRT Dynamo effort is an ongoing effort to optimize code with TensorRT using the novel `torch.compile` and `torch.export` APIs, introduced with PyTorch 2.X. Beginning from RFC #1825 and the many subsequent Dynamo RFCs, the Torch-TensorRT Dynamo integration was structured similarly to the Torch-TensorRT TorchScript integration which preceded it. The components of the Dynamo integration were mostly designed from the ground-up, using Torch utilities where possible to avoid code duplication, and mirroring much of the work done on the TorchScript path for development of converters for key operators. Below is an overview of the development of the Dynamo paths.

# Current Torch-TensorRT Dynamo Structure
### `torch.compile`
Torch compile can take almost any callable as input, including `torch.fx.GraphModule`, plain functions, and `nn.Module` types. Input handling and parsing is automatically handled by the `torch.compile` context manager and the first phase of lowering is done here by PyTorch. Specifically, graphs are formatted to have inputs and outputs which are sequences of Tensors (except in the dynamic shape case, where `torch.SymInt` can be an input, though we do not support this yet in the Torch-TensorRT backend). Additionally, complex control flow, `for`-loops, and Torch-unsupported Python syntax are automatically handled by `torch.compile`, via the Guard mechanism, which intercepts un-traceable code and runs it in Python at runtime.

## Decompositions [Lowering Phase 1]
Shared between the export and compile paths, we specify a set of decompositions which are applied by the ATen tracer (see next section). The implementation originated directly from the `core_aten_decompositions` provided by Torch, which is an evolving set of decompositions of ATen operators to other lower-level operators. Due to the evolving nature of this set, we later opted to add the user-specifiable option `enable_experimental_decompositions`. This option has two settings:
- `enable_experimental_decompositions = True`
Uses all of the `core_aten_decompositions` provided by PyTorch, _minus_ the explicitly disabled set of decompositions, _plus_ the set of [custom decompositions written by the Torch-TRT team](https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/_decompositions.py#L71-L162). See below for the set of explicitly disabled decompositions
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/_decomposition_groups.py#L178-L180
- `enable_experimental_decompositions = False`
A pre-selected set of decompositions, as implemented by Torch, _plus_ the set of [custom decompositions written by the Torch-TRT team](https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/_decompositions.py#L71-L162). See below for the set of explicitly enabled decompositions
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/_decomposition_groups.py#L13-L177

The set of explicitly enabled decompositions are a list of ATen operators whose decompositions are obtained using the `torch._decomp.get_decompositions` function.

### `torch.compile`
Lowering to ATen IR is accomplished using the `aot_export_joint_simple` function, as here:
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/backend/backends.py#L78-L86
This function takes a `torch.compile`-parsed `GraphModule` and lowers it to the ATen IR set, including applying the above decompositions.

## Graph Lowering Passes [Lowering Phase 3]
Occasionally, there are lowering passes which cannot be accomplished via a decomposition, for instance those requiring multiple operator replacement, subgraph rewriting, or constant folding. These lowering passes are completed in the third lowering phase, for which the _ordered_ passes are detailed below:
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/_aten_lowering_pass.py#L15-L25
Each pass is structured to take in a `GraphModule` and a sequence of input Tensors, and output a valid `GraphModule` which represents the graph after applying the lowering pass. A brief overview of some key current passes is below:

- `constant_fold`
Evaluates any operators containing graph constants and freezes them to graph parameters
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/constant_folding.py#L25
- `lower_linear`
Re-composes an operator (`aten.linear`) from its decomposed components
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py#L12
- `fuse_prims_broadcast`
Fuses two operators into one, simpler operator
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/fuse_prims_broadcast.py#L14
- `repair_input_as_output`
Repairs graph cases where a graph input is also a graph output
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/repair_input_as_output.py#L13
- `remove_input_alias_fixing_clones`
Enables a workaround for an issue with `aot_export_joint_simple` where clones of input Tensors had to be inserted in the GraphModule for proper tracing
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/passes/remove_input_alias_fixing_clones.py#L13

## Partitioning
There are two options for partitioning in the Dynamo paths - the "fast partitioner" and the "global partitioner". At a very high-level, the fast partitioner partitions quickly, using a greedy algorithm, but may return suboptimal partitions (extra segmentation) in certain cases. The global partitioner partitions slowly, using a global-partition-fusing algorithm, but generally returns partitions which minimize segmentation in the graph. There are a few other minor differences in these partitioners as well.

### `use_fast_partitioner=True`
Uses the `_SplitterBase` class from Torch, as show below. Returns a new `GraphModule` which has submodules `_run_on_acc_*` and `_run_on_gpu_*` for partitions to run in TensorRT or Torch, respectively. The partitions are built based on an operator support checker class shown below as well.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py#L87
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py#L28
### `use_fast_partitioner=False`
Uses the `CapabilityBasedPartitioner` class from Torch, as shown below. Returns a new `GraphModule` which has submodules `fused_*` for partitions to run in TensorRT and leaves nodes to run in Torch as-is. The partitions are built based on an operator support checker class shown below as well.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py#L21
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/partitioning/_global_partitioner.py#L130

### Conversion Process
Conversion is enabled by the `TRTInterpreter` class which runs the `torch.fx.GraphModule` line-by-line, converting each line to TensorRT via the converter implementations described below. This class selects the correct converter for each ATen operator and assigns input and output shapes and data types while building the TensorRT engine from scratch.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py#L48

The TensorRT engine itself has a few notable user-specifiable options which are `precision` (or `enabled_precisions`), `max_aux_streams`, `version_compatible`, and `optimization_level`. The documentation for each is displayed below.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py#L137-L151

### Converter Implementations
Converts each of the Torch ATen (or `prims`) operators to the TensorRT layers of the equivalent operation. All operators can be found in the file: `py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py`. Each function in the class represents the converter for one or multiple ATen ops. See below for an example of a converter. Each stacked call of `@dynamo_tensorrt_converter` represents an ATen operator which can be converted. The `@enforce_tensor_types` represents enforcement constraints on the types of Tensors which can be taken as input for a particular argument in the converter.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py#L57-L89
A detailed discussion on how to write converters can be found [here](https://pytorch.org/TensorRT/contributors/dynamo_converters.html#dynamo-converters).

## Runtime
There are two runtime options in the Dynamo paths - the Python runtime and the C++ runtime. The C++ runtime is suggested since it can be used in serialization for export, and is a more direct interface with the TensorRT API. The Python API is for usage when the C++ dependency is not present.

### `use_python_runtime = None`
Automatically selects the runtime based on the presence of the `torch_tensorrt` C++ package, with a preference for the C++ runtime.

### `use_python_runtime = True`
Uses the Python runtime shown below to run inference.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py#L14
### `use_python_runtime = False`
Uses the C++ runtime shown below to run inference.
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py#L19

**Comment by zewenli98** [comment]
@gs-olive Thanks very much for the documentation! I believe this can help developers who are not familiar with this project understand Torch-TRT as a whole. After reading it, I have some questions, from a newbie perspective. 

> ## Decompositions [Lowering Phase 1]
> Shared between the export and compile paths, we specify a set of decompositions which are applied by the ATen tracer (see next section). 

1. You mentioned `export and compile paths`. How many paths do we have? Why should we need these paths? Any difference between them? In which case which path should be used?

> * `enable_experimental_decompositions = False`
>   A pre-selected set of decompositions, as implemented by Torch, _plus_ the set of [custom decompositions written by the Torch-TRT team](https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/py/torch_tensorrt/dynamo/lowering/_decompositions.py#L71-L162). See below for the set of explicitly enabled decompositions

2. Is this set (`A pre-selected set of decompositions`) a subset of `core_aten_decompositions`? How to pre-select? based on which norm?

3. Either `enable_experimental_decompositions` is `True` or `False`, `the set of custom decompositions written by the Torch-TRT team` will be used anyway?

> ## AOT ATen Lowering [Lowering Phase 2]
> ### `torch.compile`
> Lowering to ATen IR is accomplished using the `aot_export_joint_simple` function, as here:

4. For the function name, how to understand the words: `export`, `joint`, `simple`?

> ## Partitioning

5. Why do we need `Partitioning`? What are the input and output of `Partitioning` stage? For TensorRT optimization?

**Reply by gs-olive** [reply]
Thank you for the detailed questions. To address them:
## 1. Paths
We have four paths currently:
- TorchScript
First path developed for Torch-TRT. Uses Torch's TorchScript IR, which is a lower-level deep learning language that generally uses ATen ops as well. Best suited for models needing to be exported or serialized.
- FX
Second path developed for Torch-TRT. Uses Torch's FX IR, along with a variety of different tracing options. Dynamo paths are generally preferred over FX.
- Dynamo Export
Utilizes new Dynamo ATen + AOT utilities to create an ATen graph in FX IR. Still enables serialization and export, but does not support complex Python code or graph breaks.
- `torch.compile`
General purpose acceleration using `torch.compile`, allowing for graph breaks, complex Python code, and compiling ATen graphs in much the same way as Dynamo Export. Preferable for fast compilation, benchmarking, and debugging.

## 2. Core ATen Decompositions
The preselected set of decompositions, which is used when `enable_experimental_decompositions=False`, is selected by the Torch-TRT team for best coverage of the ATen ops in TRT. It has a lot of overlap with the `core_aten_decompositions`, but it is not a subset. Specifically, `core_aten_decompositions` includes some decompositions like `_softmax`, which Torch-TRT does not use, while our preselected set includes some decompositions which `core_aten_decompositions` does not use.

## 3. Torch-TRT Decompositions
Yes, the custom Torch-TRT decompositions will be included anyway regardless of the value of `enable_experimental_decompositions`. That flag is primarily intended to control the inclusion of `core_aten_decompositions`, so if we have any custom decompositions written for Torch-TRT, these are generally considered to take priority over any other decompositions.

## 4. `aot_export_joint_simple`
This is a torch function, [as linked here](https://github.com/pytorch/pytorch/blob/44c9e4cbf0a91685729a53da362bfe648e8446bc/torch/_functorch/aot_autograd.py#L5177), and I believe `export` implies the FX graph cannot have graph breaks, `joint` means the graph can be traced both forwards and backwards (we only use the former), and `simple` could imply that the tracer does a minimal amount of graph modification.

## 5. Why partitioning?
We need partitioning because certain operators don't have implemented converters to TRT, so the partitioning phase splits the graph into subgraphs which we have converters for (to run in TRT), and subgraphs which we do not have converters for (to run in Torch).
